Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,241 @@
+#include "Spire/Ui/DropDownMenu2.hpp"
+#include <QFocusEvent>
+#include <QKeyEvent>
+#include <QScrollBar>
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  const auto BORDER_PADDING = 2;
+
+  void adjust_current_index(DropDownMenu2& menu, int amount) {
+    if(menu.count() == 0 || amount == 0) {
+      return;
+    }
+    auto last_index = menu.count() - 1;
+    auto index = [&] {
+      auto index = menu.get_current();
+      if(!index) {
+        if(amount > 0) {
+          return 0;
+        } else {
+          return last_index;
+        }
+      }
+      *index += amount;
+      if(index > last_index) {
+        return 0;
+      } else if(index < 0) {
+        return last_index;
+      }
+      return *index;
+    }();
+    menu.set_current(index);
+  }
+}
+
+DropDownMenu2::DropDownMenu2(std::vector<DropDownItem*> items,
+    QWidget* parent)
+    : DropDownWindow(true, parent),
+      m_selected_index(0),
+      m_max_displayed_items(5) {
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  auto main_widget = new QWidget(this);
```

Comment:
Rename ```main_widget``` to ```body``` in line with how ```body``` is used in HTML. Also the owner of ```body``` should be ```m_scroll_area``` as far as I can tell.

Reply:
The QScrollArea reassigns the parent with setWidget is called, but I will make it explicit here.